### DISCUSSION TITLE
Trajectory streaming interface for MotoROS2

### DISCUSSION THREAD

**Discussion by drolleigh** [Yaskawa-Global/motoros2#150] [Q&A]
System: ubuntu 20.04, ROS2 Humble, MoveIt 2

I'm working on a project where I want to implement a streaming interface for motion plans that Moveit creates in the form of a JointTrajectory message. Ultimately, the goal is to be able to accomplish [trajectory replacement ](http://wiki.ros.org/joint_trajectory_controller/UnderstandingTrajectoryReplacement), and be streaming FollowJointTrajectory points to the MotoROS2 controller.

I think this is very similar to what you were trying in [issue 66](https://github.com/Yaskawa-Global/motoros2/issues/66).

My questions are these:
1. Is there a MotoROS2 simulated controller that I can experiment not on hardware?
2. If not MotoROS2, is there advice or somewhere to look to attempt to accomplish that?

**Comment by ted-miller** [comment]
> Is there a MotoROS2 simulated controller that I can experiment not on hardware?

Unfortunately, no. There is no simulator available for MotoROS2. 

> If not MotoROS2, is there advice or somewhere to look to attempt to accomplish that?

I don't personally have any experience with it. But I believe that Gazebo offers some simulation capabilities that are compatible with ROS control architectures. Unfortunately I cannot offer any advice on how to use it.

**Comment by gavanderhoorn** [comment]
Could you clarify a bit what the non-functional requirements are in the project you mention?

I ask because the node shared in https://github.com/Yaskawa-Global/motoros2/issues/66#issuecomment-1702497412 could theoretically be extended to support trajectory replacement. The point queuing interface offered by MotoROS2 definitely supports it (as it ingests points, not trajectories) and it would be great if we could avoid duplication of effort.

It won't approach `ros(2)_control` like levels of performance, but that'd be because of limitations in the current-gen Yaskawa controllers, not MotoROS2 necessarily.

Finally: we do have a `ros(2)_control` compatible (more) real-time streaming interface [on the roadmap](https://github.com/Yaskawa-Global/motoros2/tree/161fddb1dd3a057db0571dd83ed0837c72d0cf91#provisional-roadmap), but that would require extensions to MotoROS2, and would take more time.

If you're interested in something like that, you could take a look at [ra-mtp-ntnu/motoman_hardware](https://github.com/ra-mtp-ntnu/motoman_hardware) in combination with I believe [tingelst/motoman@realtime-motion-server](https://github.com/tingelst/motoman/tree/realtime-motion-server). I've not (recently) used this, and absolutely can't guarantee fitness-for-use, YMMV. Perhaps @tingelst could comment on the status of that code / those packages.

> If not MotoROS2, is there advice or somewhere to look to attempt to accomplish that?

could you also clarify what you refer to by *that* here?

---

I've wanted to start a discussion around the "real-time position streaming interface (skipping MotoROS2's internal motion queue)" item on the roadmap with a couple of users that have expressed interest in this in the past (both here and in the context of MotoROS1). Perhaps this would be a good time.

**Reply by gavanderhoorn** [reply]
Btw:

> I'm working on a project where I want to implement a streaming interface for motion plans that Moveit creates in the form of a JointTrajectory message.

you *could* ~~probably~~ maybe also achieve this by implementing a MoveIt controller (note that's different from a `ros(2)_control` controller handle/interface which MoveIt also supports) and have it use that for trajectory execution.

We typically recommend users to configure MoveIt to use a `moveit_simple_controller_manager/MoveItSimpleControllerManager` ([here](https://github.com/Yaskawa-Global/motoros2/tree/161fddb1dd3a057db0571dd83ed0837c72d0cf91?tab=readme-ov-file#with-moveit)) which is essentially just an action client to the `FollowJointTrajectory` server MotoROS2 exposes. But it would technically be possible to implement your own which directly uses the various point queuing services (such as [queue_traj_point](https://github.com/Yaskawa-Global/motoros2/blob/161fddb1dd3a057db0571dd83ed0837c72d0cf91/doc/ros_api.md#queue_traj_point)) to stream trajectories to MotoROS2.

It would be specific to MoveIt, and it would still need trajectory replacement functionality, but that could potentially be imported from the `ros(2)_control` joint trajectory controller.

**Comment by gavanderhoorn** [comment]
@drolleigh: have you been able to make any progress?
